Decide inline-worthiness based on a more nuanced cost model

timholy opened the pull request:
In #22202 I noted a disadvantage of our current inlining heuristics: that we penalize verbosity even when two bits of code do the same thing. I decided to explore the feasibility of replacing our cost model with something based a little more closely on CPU cycles, with the idea that we should inline things that are obviously cheaper than a function call. I decided to be a bit generous and set a default threshold of 200 "cycles", and assigned costs to operations loosely inspired by [this chart](http://ithare.com/infographics-operation-costs-in-cpu-clock-cycles/).

This is a horribly rough approximation: for example, there's a very large difference between success and failure when it comes to branch-prediction, and it might be reasonable to guess that the branch in
```julia
x < zero(x) && throw(DomainError())
```
is much less likely to be hit than
```julia
if x < rand()
    # option 1
else
    # option 2
end
```
Such subtleties (not to mention cache prefetch, etc) are far beyond the simple approach here. In general I tried to aim for an optimistic model, but worked in some amount of penalty for the possibility that this wouldn't hold, with the amount depending entirely on how I was feeling at that particular moment :smile:.

It's also worth pointing out that I've done nothing whatsoever to test how well this works. My entire goal was to get julia to build.

This still has debugging statements in the submitted code because I can't imagine that this isn't going to need at least a teensy bit of revision :wink:.

timholy commented:
Just realized there's a really big hole in this: a `for` or `while` loop will only get counted as if its body is executed once. Probably it would be best to just not inline functions with such loops. I don't see a "loop detector" presently, but perhaps I'm just missing it.

timholy commented:
OK, I've taken this further; a modified version of this (that had some debugging code) locally passed all tests but one. For demonstration purposes, I've also pushed a branch, `teh/disable_inline_macros`, that gets rid of all forced-inlining in Base; the point of this is to force the inliner to "stand on its own" and see what happens. I then rebased this PR on that branch. My original intention was to then run nanosoldier to compare the differences, but from my local testing there are so many differences that it's a bit overwhelming.

The bottom line seems pretty simple, though. Behold:
![inline_perf](https://user-images.githubusercontent.com/1525481/26852137-e6730dec-4ad2-11e7-8b30-a4f0100f8657.png)

So while there are a few regressions, they are both outnumbered and outweighed by the improvements.

As one example, `map(abs, t)` where `t = ntuple(identity, n)` inlines (on its own) up to `n=3` on master but `n=14` on this branch; a more expensive function, `map(sqrt, t)`, inlines up to `n=5`.

Now, one concern might be that this simply inlines everything, and that's cheating. What's the best way to assess such concerns? I checked the "resident size" in `top`, and compared to 0.6 (I didn't feel like rebuilding master) the size was pretty comparable.

yuyichao commented:
Am I reading it correctly that there's one case that shows 10^6 speed up?

> Now, one concern might be that this simply inlines everything, and that's cheating.

Well, then all optimizing compilers have been cheating all alone.... I'm pretty sure most C/C++ compiler do much more aggressive inlining than we do so I wouldn't call it cheating. The important thing is to measure the overhead. Maybe measuring the time to compile sysimg, or some other standard precompilation test cases. You can also measure the total size in inferenced IR record a statistic of non-void expression in inference after finishing each functions and if you want more low level info you can also check `size -A` of sysimg and record the JIT code size by adding a counter in `RTDyldMemoryManagerJL::allocateCodeSection` in `cgmemmgr.cpp`.

JeffBezanson commented:
Wow, that's really cool. If you re-enable manual inlining declarations, do the regressions go away?

timholy commented:
> Am I reading it correctly that there's one case that shows 10^6 speed up?

Yes, that's basically right; several others cases get better than 10^4 speed up. Or rather, "not slowed down"; I am sure that our performance annotations take care of all the egregious cases. I very much doubt we'll see much improvement compared to "real" master from this, but of course it's worth testing.

> Maybe measuring the time to compile sysimg

Great idea.

| task | master | PR |
|:-------:|:----------:|:------:|
| build from scratch (incl. `inference.jl`) | 2min 46s | 3min 6s |
| build `base/precompile.jl` | 1min 36s | 1min 50s |

So there's definitely a slowdown with this. Would be interested in people's thoughts on how bad this is.

> If you re-enable manual inlining declarations, do the regressions go away?

Will test. But right now I'm testing how this fares against "real" master, with all of its performance annotations; I'm curious to know if we could scale back on our usage of `@inline`.

yuyichao commented:
I believe we also have a `ENABLE_TIMING` option in `options.h` which can give you a breakdown of the timing. That timing different doesn't sound too bad to me though compile time is also not the thing I care about the most....

timholy commented:
Here's a comparison against master with all of its `@inline` annotations:

![inline_vs_master](https://user-images.githubusercontent.com/1525481/26860267-b39fbebc-4b02-11e7-9566-c03d8319beb2.png)

Definitely worse on balance, but better than I expected: to my eye, this branch *without* performance annotations is closer to master *with* annotations than it is to master with the annotations stripped out.

Obviously some of the regressions are pretty severe, but I've done literally no performance tuning here. Probably worth looking into some of the regressions by hand and seeing what I can come up with...in particular, the ones that have ~10-100ns of runtime on master seem likely to be just me assigning the wrong cost and it therefore failing to inline a pretty simple function.

timholy committed:
Protect some `throw`s with atsign-noinline

timholy committed:
Force specialization in number of varargs in key places

This greatly reduces the cost of failure-to-inline

quinnj commented:
I think it'd be really cool to expose the internal cost model here via profiling; I think it'd be a neat alternative to time/allocations/samples and another good signal in terms of identifying code performance bottlenecks and the relative "cost" of various code paths. Sometimes you don't realize the relative costs of various parts of a workflow or have some kind of type instability, all of which seems like would stick out with seeing this kind of data.

timholy commented:
@quinnj, the cost model is almost certainly not accurate enough to expose to users; it doesn't take into account looping, and the cost of non-inferrability is handled crudely. 

However, it can be useful to understand why a particular function didn't get inlined. I'll document how to call `statement_cost` for a function body.

timholy commented:
So, I'm thinking of putting this on the back burner until #21888 merges; I'm seeing quite a few issues that I hope get resolved by better GC marking. The last commit here is just one example; I'm also seeing a few big performance regressions that seem to occur because we allocate when we don't need to; a good example is `mean(rand(10^3))`, where I'm getting allocations due to unnecessarily boxing the [`blksize`](https://github.com/JuliaLang/julia/blob/a52a78d3f8a1e12b9586dab8c5d1b3ed7c626126/base/reduce.jl#L177) argument. I'm pretty sure this is not caused by my changes here, but it does get triggered by these changes.

Of course the question is whether its benefits on LLVM 3.9 will fix these issues, or whether it will have to wait until we move to LLVM 5.

timholy commented:
Speaking of bugs triggered by this, https://travis-ci.org/JuliaLang/julia/builds/240808331 hit an assertion in `gf.c`. (It was the `Int` version that was a subtype of the `Int...` version of that function.) I worked around it by declaring that argument as `Vararg{Int,N}`, but it suggests a bug elsewhere.

timholy committed:
Work around penalty from splatting an Integer

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22210#issuecomment-309283556) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/77af818db5d74c448f55c75b4368653181a80b0d/37a9727_vs_e9f2072/report.md). cc @jrevels

KristofferC commented:
Cool stuff!

timholy commented:
That is a pretty darn satisfying number of green checkboxes on that nanosoldier run. In graphical format, here are my local results:
![inlining](https://user-images.githubusercontent.com/1525481/27263533-bdca1fea-5430-11e7-93ff-5b352709c511.png)

I still have a couple of errors to fix, though. This has the habit of triggering assertions in the C code when building with `FORCE_ASSERTIONS=1` (which is how Travis builds it). The latest is 
```
julia: /home/tim/src/julia-master/src/typemap.c:843: jl_typemap_level_assoc_exact: Assertion `(((jl_value_t*)(((jl_taggedvalue_t*)((char*)(ty) - sizeof(jl_taggedvalue_t)))->header & ~(uintptr_t)15))==(jl_value_t*)(jl_datatype_type))' failed.

signal (6): Aborted
while loading /home/tim/src/julia-master/test/linalg/diagonal.jl, in expression starting on line 10
```
If anyone has any ideas, I'm all ears.

vtjnash commented:
This change also makes the sysimg about 50% larger and generates about 50% more code (see `size -A` results in the CI logs). That would generally seems to indicate this is just being much too aggressive at inlining, and not necessarily much more intelligent.

timholy commented:
That's an excellent point, and I bet you're right that it's a big part of the advantage of this PR. But rather than speculate, we can test it concretely by tuning this algorithm to have similar system size and then comparing performance, both with and without all of our current performance annotations (`@inline` and `@noinline` statements). To make it more fun, I'll "bet" that the two options may be fairly similar given our current (carefully-tuned via experience) suite of performance annotations, but that if we discard the performance annotations then this one is still going to win handily. But that's an emiprical question, so let's just find out.

See also https://github.com/JuliaLang/julia/pull/22210#issuecomment-306623682 for what appears to be a suggestion that we should become more aggressive. But, as you say, that's really an orthogonal question because it's applicable to either algorithm.

timholy commented:
I will wait for #22481, however.

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

cool that the cost model is exposed so visibly. is there a summary somewhere of how the modeled cost is used? is it strictly for inlining (what's the criterion?) or are there other things it could be used for?

timholy commented:
Just pushed a revision that included some suggestions from @vtjnash at JuliaCon. This passes tests locally, and with these settings the system image is only 7% bigger than without the last two commits. I'm still running the performance tests locally, but earlier versions of this still looked better than master. I'll say more when I have new numbers. The important point is that I think this is finished, and is **ready for a serious review** from someone who knows inference better than me.

There's one line that I had to "cheat" on. It turns out that being strict about penalizing calls which appear to be non-inferrable runs up against problems when it comes to tuple-destructuring because of the way we express tuple iteration. Here's a test file that illustrates the problem:
```julia
function destructa(a, b, c)
    s = zero(eltype(b))
    @assert length(a) == length(b) == length(c)
    for i = 1:length(a)
        va, vb, vc = destructt(a, b, c, i)
        s += vb
    end
    s
end

function destructt(a, b, c, i)
    @inbounds va = a[i]
    @inbounds vb = b[i]
    @inbounds vc = c[i]
    va, vb, vc
end

a = rand(10^6)
b = rand(Int, 10^6)
c = rand(Int, 10^6)

destructa(a, b, c)
@profile destructa(a, b, c)
Profile.print()
```
On current master this collects no samples because it's blazingly fast. On this branch, if I were to uncomment [this line](https://github.com/JuliaLang/julia/blob/2b537b9988567dd8e5b9c5363f516f1a35a96c8b/base/inference.jl#L4651), it would be much slower. This seems to arise because `start`, `next`, and `done` use an integer index for tuple iteration, and for heterogeneous tuples this makes the result nominally non-inferrable. I'm guessing we work around that somehow, but I'm a bit unclear on how that happens.

timholy commented:
@tkelman I'll write up a devdocs on this once it's been reviewed.

timholy committed:
More inlining prevention

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

Locally here's what I'm seeing:
![inline](https://user-images.githubusercontent.com/1525481/27761001-c71cbca4-5e19-11e7-9547-eba8abc363c4.png)
(System size is only 7% larger.)

And if I disable `@inline` annotations on both this branch and master, I get this:
![inline_noannotations](https://user-images.githubusercontent.com/1525481/27761007-d59ea440-5e19-11e7-94c1-def90b8c1931.png)

I've taken a look at some of the larger performance gains on a slightly outdated version of this PR. In every case I checked, it's because LLVM is getting to do more analysis of the code. There were 4 tests that were more than 100x faster (3 bitshift tests and 1 `sub2ind` test), and in all cases LLVM reduced the operation to a constant. (You might call that cheating.) But more excitingly, quite a few tests (28) got more than 10x faster, and several involve SubArrays; in those cases it was eliding the actual creation of the view, reducing it to a direct operation on the underlying array. While we may need to add some `@noinline`s to BaseBenchmarks if we want to keep those tests doing what they are now, I'll point out that this is *exactly* the kind of optimization we *want* to achieve (#14955), so I think it's a fair case.

timholy commented:
Finally, I should confess that I'm running out of time and getting to the point where I may need to disappear from core Julia development for a couple of months. If this can be reviewed soon it would be a big help.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22210#issuecomment-312423053) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/830a3512f5c7fa6239655709de1ab6964d1a382e/ae3f0b9_vs_b5a61db/report.md). cc @jrevels

tkelman commented on base/inference.jl:
```diff
@@ -427,104 +439,103 @@ function fptosi_tfunc(x::ANY)
-add_tfunc(trunc_llvm, 1, 1, math_tfunc)
-add_tfunc(rint_llvm, 1, 1, math_tfunc)
-add_tfunc(sqrt_llvm, 1, 1, math_tfunc)
-add_tfunc(sqrt_llvm_fast, 1, 1, math_tfunc)
```
where'd this go?

tkelman commented on base/reshapedarray.jl:
```diff
@@ -223,21 +225,17 @@ end
     @inbounds parent(A)[ind2sub_rs(A.mi, sub2ind(size(A), indexes...))...] = val
     val
 end
-@inline function _unsafe_setindex!(A::ReshapedArrayLF, val, indexes::Int...)
```
why is this no longer necessary?

timholy replied:
Generic array handling infrastructure will call `sub2ind` for us. (This was just a "stale" method that I happened to notice while working on this.)

Sacha0 commented on base/abstractarray.jl:
```diff
@@ -1010,7 +1010,7 @@ _setindex!(::IndexStyle, A::AbstractArray, v, I...) =
 _setindex!(::IndexLinear, A::AbstractArray, v, i::Int) = (@_propagate_inbounds_meta; setindex!(A, v, i))
 _setindex!(::IndexLinear, A::AbstractArray, v) = (@_propagate_inbounds_meta; setindex!(A, v, _to_linear_index(A)))
-function _setindex!(::IndexLinear, A::AbstractArray, v, I::Int...)
+function _setindex!(::IndexLinear, A::AbstractArray, v, I::Vararg{Int,M}) where {M}
```
In the related long-form `_getindex` method definitions earlier in this file, you omit the curly braces around the terminating `M`. For consistency, perhaps do the same here?

Sacha0 commented on base/array.jl:
```diff
@@ -147,23 +147,23 @@ function reinterpret(::Type{T}, a::Array{S}) where T where S
-    end
-    if !isbits(S)
-        throw(ArgumentError("cannot reinterpret Array{$(S)} to ::Type{Array{$(T)}}, type $(S) is not a bits type"))
+    function throwbits{S,T,U}(::Type{S}, ::Type{T}, ::Type{U})
```
Perhaps use `where` syntax?

Sacha0 commented on base/reshapedarray.jl:
```diff
@@ -96,21 +96,20 @@ reshape(parent::AbstractArray, dims::Int...) = reshape(parent, dims)
 @inline function _reshape_uncolon(A, dims)
+    @noinline throw1(dims) = throw(DimensionMismatch("new dimensions $(dims) may have at most one omitted dimension specified by Colon()"))
+    @noinline throw2(A, dims) =
+        throw(DimensionMismatch("array size $(length(A)) must be divisible by the product of the new dimensions $dims"))
```
Perhaps split the definitions of `throw1` and `throw2` over additional lines? (They both considerably exceed the suggested line length.)
```julia
    @noinline throw1(dims) = throw(DimensionMismatch(string("new dimensions $(dims) ",
        "may have at most one omitted dimension specified by Colon()")))
    @noinline throw2(A, dims) = throw(DimensionMismatch(string("array size $(length(A)) ",
        "must be divisible by the product of the new dimensions $dims")))
```

Sacha0 commented:
Out of curiosity, what was the general impact of automagic inlining of the various `throw(...)` calls that this pull request replaces with calls to short `@noinline` methods? `foo == bar || throw(...)` being a common pattern, that potentially becoming a performance trap might be suboptimal? Thanks!

Sacha0 commented on base/inference.jl:
```diff
@@ -14,6 +14,9 @@ struct InferenceParams
     inlining::Bool
+    inline_cost_threshold::Int  # number of CPU cycles beyond which it's not worth inlining
+    inline_nonleaf_penalty::Int # penalty for dynamic dispatch
+    inline_tupleret_bonus::Int  # extra willlingness to inline for non-isbits tuple return types
```
"willlingness" sounds particularly willing :).

Sacha0 commented on base/inference.jl:
```diff
@@ -374,17 +381,22 @@ isconstType(t::ANY) = isType(t) && (isleaftype(t.parameters[1]) || t.parameters[
+function add_tfunc(f::IntrinsicFunction, minarg::Int, maxarg::Int, tfunc::ANY, cost::Int)
+    idx = reinterpret(Int32,f)+1
+    t_ifunc[idx] = (minarg, maxarg, tfunc)
+    t_ifunc_cost[idx] = cost
```
I imagine nothing relies upon this method's (changed) return value?

timholy replied:
Good catch. Doesn't look like it, though.

Sacha0 commented on base/inference.jl:
```diff
@@ -427,104 +439,104 @@ function fptosi_tfunc(x::ANY)
-add_tfunc(sqrt_llvm, 1, 1, math_tfunc)
-add_tfunc(sqrt_llvm_fast, 1, 1, math_tfunc)
+add_tfunc(abs_float, 1, 1, math_tfunc, 5)
+add_tfunc(copysign_float, 2, 2, math_tfunc, 5)
```
(TIL floating point `abs` and `copysign` are surprisingly expensive!)

timholy commented:
Thanks for the reviews, @tkelman and @Sacha0!

> what was the general impact of automagic inlining of the various throw(...) calls

Great question, and I didn't know the answer anymore myself (might have changed as this PR evolved), so worth testing explicitly:

On 0.6:
```julia
julia> function foo1(A)
           size(A, 2) > 5 || throw(DimensionMismatch("A should have dimension 2 > 5, got $(size(A))"))
           A[1, 6]
       end
foo1 (generic function with 1 method)

julia> function foo2(A)
           @noinline throw1(A) = throw(DimensionMismatch("A should have dimension 2 > 5, got $(size(A))"))
           size(A, 2) > 5 || throw1(A)
           A[1, 6]
       end
foo2 (generic function with 1 method)

julia> using BenchmarkTools

julia> A = rand(10,10);

julia> @benchmark foo1(A)
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     22.167 ns (0.00% GC)
  median time:      23.652 ns (0.00% GC)
  mean time:        26.165 ns (2.54% GC)
  maximum time:     3.395 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     996

julia> @benchmark foo2(A)
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     20.814 ns (0.00% GC)
  median time:      22.010 ns (0.00% GC)
  mean time:        24.757 ns (2.59% GC)
  maximum time:     1.073 μs (90.81% GC)
  --------------
  samples:          10000
  evals/sample:     997
```

On this branch:
```julia
julia> @benchmark foo1(A)
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     25.638 ns (0.00% GC)
  median time:      26.342 ns (0.00% GC)
  mean time:        32.929 ns (16.16% GC)
  maximum time:     37.796 μs (99.92% GC)
  --------------
  samples:          10000
  evals/sample:     995

julia> @benchmark foo2(A)
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     20.813 ns (0.00% GC)
  median time:      21.989 ns (0.00% GC)
  mean time:        28.990 ns (20.48% GC)
  maximum time:     42.903 μs (99.92% GC)
  --------------
  samples:          10000
  evals/sample:     997
```

Seems to be a very slight difference. I seem to remember it was much bigger in an earlier iteration of this, so I suspect that I've fixed it "generically" and may not need that commit anymore. Should I drop it? Or is the small improvement worth it?

timholy commented:
Since @vtjnash looked at this briefly at JuliaCon, in the absence of further comments would it be reasonable to merge this by Saturday (or so)? If someone needs more time, just say the word.

tkelman commented on base/reshapedarray.jl:
```diff
@@ -96,9 +96,13 @@ reshape(parent::AbstractArray, dims::Int...) = reshape(parent, dims)
 @inline function _reshape_uncolon(A, dims)
-    @noinline throw1(dims) = throw(DimensionMismatch("new dimensions $(dims) may have at most one omitted dimension specified by Colon()"))
+    @noinline throw1(dims) = throw(DimensionMismatch(
+"new dimensions $(dims) may have at most one omitted dimension specified by `Colon()`"
```
I'd break the string, this looks odd to me

Sacha0 commented:
> Seems to be a very slight difference. I seem to remember it was much bigger in an earlier iteration of this, so I suspect that I've fixed it "generically" and may not need that commit anymore. Should I drop it? Or is the small improvement worth it?

Perhaps rerun those benchmarks interpolating `A` (e.g. `foo($A)`)? :)

timholy commented:
> Perhaps rerun those benchmarks interpolating A (e.g. foo($A))? :)

Duh, right. Corrected above.

Sacha0 commented on base/inference.jl:
```diff
@@ -3245,11 +3257,12 @@ end
-    cost = 1000
+    cost_threshold = params.inline_cost_threshold
     if m.module === _topmod(m.module)
+        # a few functions get special treatment
```
Perhaps `throw` should be treated specially as well, to avoid the necessity of wrapping `throw`s into `@noinline`'d methods?

timholy replied:
I'm not entirely clear on how one would do that automatically. It's also not a very big performance hit. I think I made those changes when this was at an earlier stage and it caused a bigger performance hit. In particular [this line](https://github.com/JuliaLang/julia/blob/abcdc77b58a6f58f9cf376e820aefb7e445ee7fd/base/inference.jl#L4666), which  runs for any `:call` expr that's not an Intrinsic/Builtin which means that it's a dynamic dispatch---if it were statically known it would be an `:invoke`---is probably what mostly fixes it.

timholy committed:
Decide inline-worthiness based on a more nuanced cost model

This switches to a model in which we estimate the runtime of the
function; if it's fast, then we should inline it. The estimate of
runtime is extremely crude, and doesn't even take loops into account.

timholy commented:
If this makes it through CI, I'll merge in the morning.

vchuravy commented:
Appveyor is failing due to libgit2-online. Do all the intermediate commits pass tests as well?

timholy commented:
Yes, they do.

KristofferC commented:
Yay!

martinholters commented:
Hm, looks like this makes the ACME.jl tests [crash](https://travis-ci.org/HSU-ANT/ACME.jl/jobs/252309301). Not sure if the problem was introduced here or just uncovered by different inlining. I haven't really gotten anywhere investigating the issue, but somewhat up the call stack of a sightly reduced case, I found this:
```
#16 0x00007ffff7574ed4 in jl_call_fptr_internal (fptr=0x7fffffffb940, meth=0x7fffe999ec90, args=0x7fffffffbb48, nargs=2)
    at ./src/julia_internal.h:369
369	        return fptr->fptr1(args[0], &args[1], nargs-1);
(gdb) call jl_(args[0])
getfield(Base, Symbol("##494#495")){Base.MethodError}(ex=Base.MethodError(f=typeof(Base.:(*))(), args=<?#0x7fffe906f230::Core.Inference.InferenceParams(world=0x0000000000005677, inlining=true, inline_cost_threshold=100, inline_nonleaf_penalty=1000, inline_tupleret_bonus=400, MAX_METHODS=4, MAX_TUPLETYPE_LEN=15, MAX_TUPLE_DEPTH=4, MAX_TUPLE_SPLAT=16, MAX_UNION_SPLITTING=4, MAX_APPLY_UNION_ENUM=8)>, world=0x0000000000005677))
```
Note that the _type_ of (inner) args is an _instance_ of `InferenceParams`. Of course, this could also just be a random artifact of some memory corruption. Does anyone have any ideas or has seen anything similar?

timholy commented:
As long as it doesn't crash in `inline_worthy` or `statement_cost`, there's a sense in which this can't introduce *new* bugs: the only thing it changes is the value of a single boolean flag, the decision to inline or not. It doesn't change anything about how that inlining is performed.

However, it's perfectly capable of exposing old bugs. Indeed it's already done so: at least #22516, and (at an intermediate stage) what is probably the same bug as #22592.

I've not seen that one before, though.

martinholters commented:
Seems indeed to be a missing GC root again, ref. https://github.com/JuliaLang/julia/issues/22770#issuecomment-314689818.

tkelman commented:
> I'll write up a devdocs on this once it's been reviewed.

Sometime later then if you have time?
